README.md
# Predis ServiceProvider #

[![Latest Stable Version](https://poser.pugx.org/predis/service-provider/v/stable.png)](https://packagist.org/packages/predis/service-provider)
[![Total Downloads](https://poser.pugx.org/predis/service-provider/downloads.png)](https://packagist.org/packages/predis/service-provider)
[![License](https://poser.pugx.org/predis/service-provider/license.svg)](https://packagist.org/packages/predis/service-provider)
[![Build Status](https://travis-ci.org/nrk/PredisServiceProvider.svg?branch=master)](https://travis-ci.org/nrk/PredisServiceProvider)
[![HHVM Status](http://hhvm.h4cc.de/badge/predis/service-provider.png)](http://hhvm.h4cc.de/package/predis/service-provider)

This service provider for [Silex](http://silex-project.org/) allows developers to easily configure
and expose [Predis](http://github.com/nrk/predis) enabling them to use [Redis](http://redis.io) in
their applications.


### Getting started ###

Supposing that the skeleton of your application is ready, you simply need to register this service
provider by specifying the parameters and options needed to access Redis:

```php
$app->register(new Predis\Silex\ClientServiceProvider(), [
    'predis.parameters' => 'tcp://127.0.0.1:6379',
    'predis.options'    => [
        'prefix'  => 'silex:',
        'profile' => '3.0',
    ],
]);
```

This will register one instance of `Predis\Client` accessible from anywhere in your application by
using `$app['predis']`. Both `predis.parameters` and `predis.options` are optional and they accept
the same values accepted by the constructor of `Predis\Client` (see the documentation of Predis).

Certain applications might need more than one client to reach different servers or configured with
different options. In such cases you must use `Predis\Silex\ClientsServiceProvider` providing a list
of clients with their own parameters and options using `predis.clients`:

```php
$app->register(new Predis\Silex\ClientsServiceProvider(), [
    'predis.clients' => [
        'client1' => 'tcp://127.0.0.1:6379',
        'client2' => [
            'host' => '127.0.0.1',
            'port' => 6380,
        ],
        'client3' => [
            'parameters' => 'tcp://127.0.0.1:6381',
            'options' => [
                'profile' => 'dev',
                'prefix'  => 'silex:',
            ],
        ],
    ),
]);
```

Clients will be exposed to your application using `$app['predis'][$alias]` where `$alias` is the key
used to populate the items of `predis.clients`. Optionally, it is possible to set a default client
by specifying its alias in `predis.default_client` making it accessible simply by invoking methods
of `Predis\Client` directly against `$app['predis']`. Client instances are lazily initialized upon
the first access.

__NOTE__: this is not the same as using a cluster of nodes or replication as it will only create and
set up independent client instances. Cluster and replication thus work with both single and multiple
client configurations, you just need to provide the needed parameters and options for each instance
of `Predis\Client`.

You can find more details on how to use this provider in the `examples` directory or the test suite.

### Reporting bugs and contributing code ###

Contributions are highly appreciated either in the form of pull requests for new features, bug fixes
or just bug reports. We only ask you to adhere to a [basic set of rules](CONTRIBUTING.md) before
submitting your changes or filing bugs on the issue tracker to make it easier for everyone to stay
consistent while working on the project.


### Project links ###

- [Source code](http://github.com/nrk/PredisServiceProvider)
- [Issue tracker](http://github.com/nrk/PredisServiceProvider/issues)


### Author ###

- [Daniele Alessandri](mailto:suppakilla@gmail.com)
  ([github](http://github.com/nrk))
  ([twitter](http://twitter.com/JoL1hAHN))


### Contributors ###

- Jérôme Macias ([github](http://github.com/jeromemacias))


### License ###

The code for Predis ServiceProvider is distributed under the terms of the [MIT license](LICENSE).


CHANGELOG
v1.0.0 (2014-08-xx)
  * Switched to PSR-4 for autoloading.

v0.4.2 (2013-10-18)
  * Fix Silex dependency to allow stable and dev versions.

v0.4.1 (2013-03-19)
  * Added `predis.client_constructor` to easily override the actual instance
    of `Predis\ClientInterface` returned by the provider when creating a new
    client object.

  * Minimum required version of Predis is now v0.8.3.

v0.4.0 (2013-02-03)
  * The service provider now requires Predis v0.8.

  * `Predis\Silex\PredisServiceProvider` is now used only for single-client
    configurations while the new `Predis\Silex\MultiPredisServiceProvider`
    can be used to configure multiple clients. This is a breaking change from
    previous versions in which both configurations were handled by the same
    provider class.

  * Due to the nature of the above mentioned change, it is no more possible
    to define a default client when using a multiple-clients configuration.

v0.3.0 (2012-07-14)
  * Fixed service provider configuration after the recent addition of the
    `boot()` method in Silex.

  * Removed the `class_path` option, Silex removed its autoloader service and
    now relies on Composer.

v0.2.4 (2012-06-01)
  * Implemented an empty Silex\ServiceProviderInterface::boot() needed after
    recent changes in Silex.

v0.2.3 (2011-12-11)
  * There are no actual changes in the code of this release, aside from a bump
    in the required version of Predis in the composer.json file to target the
    tagged release of v0.7.0,

v0.2.2 (2011-11-13)
  * Added the ability to specify multiple clients using 'predis.clients' with
    an array of connection parameters and client options for each client.

v0.2.1 (2011-10-02)
  * Added support for Composer to manage installation and dependencies.

v0.2.0 (2011-10-02)
  * Renamed PredisExtension to PredisServiceProvider after internal changes
    in the naming of some classes and interfaces of Silex.

v0.1.0 (2011-05-07)
  * Initial release.


examples/multiple.php
<?php

require __DIR__.'/../vendor/autoload.php';

$app = new Silex\Application();

$app->register(new Predis\Silex\ClientsServiceProvider(), array(
    'predis.clients' => array(
        'first' => 'tcp://127.0.0.1:6379',
        'second' => array(
            'host' => '127.0.0.1',
            'port' => 6380,
        ),
        'third' => array(
            'parameters' => array(
                'host' => '127.0.0.1',
                'port' => 6381,
            ),
            'options' => array(
                'profile' => '2.2',
                'prefix' => 2.4,
            ),
        ),
    ),
    'predis.default_client' => 'first',
    'predis.default_parameters' => array(
        'timeout' => 2.0,
        'read_write_timeout' => 2.0,
    ),
    'predis.default_options' => array(
        'connections' => array(
            'tcp'  => 'Predis\Connection\PhpiredisConnection',
            'unix' => 'Predis\Connection\PhpiredisConnection',
        ),
    ),
));

/** routes **/

$app->get('/', function () use ($app) {
    $default = var_export($app['predis']->info('server'), true);

    $first = var_export($app['predis']['first']->info('server'), true);
    $second = var_export($app['predis']['second']->info('server'), true);
    $third = var_export($app['predis']['third']->info('server'), true);

    return "$default<br/>\n$first<br/>\n$second<br/>\n$third<br/>\n";
});

/** run application **/
$app->run();
